Post: Why all this fuss about eventual consistency? It's everywhere! (#321)
mauroservienti/milestone.github.io: _posts/fuss-about-eventual-consistency.md

  - If the supplier doesn't exist in the system, they are presented with a screen to add a new supplier.

Once the step is completed, they return to the product creation screen and expect to use the recently created supplier immediately. That's a scenario when users need to "read their own writes."

In contrast, this blog is eventually consistent, and that's just fine:

- I write a markdown document.
- Push changes to GitHub.
- GitHub pages schedule an Action to regenerate the blog.
- After regeneration, an Action step publishes the new content to GitHub pages.

That's eventual consistency at play. It can take a few minutes to complete, and that's fine.

From the end-user perspective, that's the same that happens when creating a product in an e-commerce system. They don't know the creation process is happening, and they're not waiting.

Readers and writers are different personas. The reader is unaware of when the writer pushes changes and thus is not eagerly waiting for the newly created article or product to be available. Who writes differs from who reads.

Eventual consistency causes headaches, primarily if not only in the "read your own writes" scenario.

> ## A note: Eventual consistency is with us

Reviewer: ```suggestion
## A note: Eventual consistency is with us...
```

> And has always been unless pessimistic locking was in place. Let's look at another example: in the previously mentioned e-commerce, there is a strange rule, also known as invariant—[which is evil](https://milestone.topics.it/2021/05/05/tales-of-a-reservation.html), stating that customers cannot buy any out-of-stock product. The only possible way to enforce the rule is to serialize requests to the database using the highest possible transaction level. If that's not the case, there's always a chance of [phantom reads](https://sqlsolutionsgroup.com/dodging-phantom-reads/) leading to more than one customer ordering the last-in-stock [lightsaber](https://en.wikipedia.org/wiki/Lightsaber).

Reviewer: ```suggestion
...and has always been unless pessimistic locking was in place. Let's look at another example: in the previously mentioned e-commerce scenario, there is a strange rule, also known as invariant—[which is evil](https://milestone.topics.it/2021/05/05/tales-of-a-reservation.html), stating that customers cannot buy any out-of-stock product. The only possible way to enforce the rule is to serialize requests to the database using the highest possible transaction level. If that's not the case, there's always a chance of [phantom reads](https://sqlsolutionsgroup.com/dodging-phantom-reads/) leading to more than one customer ordering the last-in-stock [lightsaber](https://en.wikipedia.org/wiki/Lightsaber).
```

From the programming model perspective, it's mostly about complete consistency. We want all of them to be written at once without incurring partial writes. At the same time, we want all of them to be available for readers when they are all written to avoid phantom reads. For example, we don't wish cart items to be available for reads while we still write the owning shopping cart.

## Distributed transactions, DCOM and CORBA

But then distributed transactions came, and we got carried away. In their simplest form, distributed transactions allow a transaction between two database tables defined in different databases on different computers.

We did not stop there. Different resources could be enlisted in a distributed transaction, like a database table, a message queue, and a remote service invocation. Can you imagine all those heterogeneous things in the same transaction?

We even went further with protocols like DCOM or CORBA, presenting to consumers a set of distributed object instances as if they were a single instance. All of them coordinated by transactions.

What could possibly go wrong?

## The cloud changed everything

> Transactions were so engraved that for some, it was hard, if ever possible, to think about a world without transactions. But "recently," something changed. The cloud became an affordable and available option to the masses. And the cloud, primarily for the vendors' benefit, decided that (distributed) transactions were to be banned in disgrace. The underlying reason is that infrastructure in the cloud is flaky in nature, and transaction protocols were thought and designed under the assumption that [the network is reliable](https://particular.net/blog/the-network-is-reliable), the number one fallacy of distributed computing.

Reviewer: ```suggestion
Transactions were so engraved that for some, it was hard, even impossible, to think about a world without transactions. But "recently," something changed. The cloud became an affordable and available option to the masses. And the cloud, primarily for the vendors' benefit, decided that (distributed) transactions were to be banned in disgrace. The underlying reason is that infrastructure in the cloud is flaky in nature, and transaction protocols were thought and designed under the assumption that [the network is reliable](https://particular.net/blog/the-network-is-reliable), the number one fallacy of distributed computing.
```

## A new hope (cit.)

The cloud also came with a neat solution to the problem: messaging, message queues, and message brokers. Instead of having so many heterogeneous resources in a single transaction, developers can use reliable messaging infrastructure to send requests to other services.

Messaging allows splitting significant processes into smaller chunks and connecting them using messages. A bit of a process happens there, and then a message gets sent to a different node to continue with another bit. And so on until the process is completed. 

Messaging is a way to trade consistency for availability.

> The proposed exchange brings eventual consistency into the mix. Because what previously was a monolithic process governed by transactions locking the requestor and forcing them to wait, it's now a set of smaller steps independent of each other with no control over consistency, making it hard for requestors to wait for completion.

Reviewer: ```suggestion
The proposed exchange brings eventual consistency into the mix. Because what previously was a monolithic process governed by transactions locking the requestor and forcing them to wait, is now a set of smaller steps independent of each other with no control over consistency, making it hard for requestors to wait for completion.
```

## So what?

> Not all hope is lost, and there are a few things we can do to solve, alleviate, or workaround any eventual consistency-related issue.

Reviewer: ```suggestion
Not all hope is lost, and there are a few things we can do to solve, alleviate, or work around any eventual consistency-related issues.
```

> ### The one architecture to rule'em all fallacy

Reviewer: ```suggestion
### The "one architecture to rule them all" fallacy
```
